core/src/main/kotlin/xtdb/api/storage/SimulatedObjectStore.kt

```diff
@@ -0,0 +1,88 @@
+package xtdb.storage
+
+import xtdb.api.storage.ObjectStore
+import xtdb.multipart.IMultipartUpload
+import xtdb.multipart.SupportsMultipart
+import java.nio.ByteBuffer
+import java.nio.file.Path
+import java.util.concurrent.CompletableFuture
+
+enum class StoreOperation {
+    PUT, UPLOAD, COMPLETE, ABORT
+}
+
+class SimulatedObjectStore(
+    val calls: MutableList<StoreOperation> = mutableListOf(),
+    val buffers: MutableMap<Path, ByteBuffer> = mutableMapOf()
+) : ObjectStore, SupportsMultipart {
+
+    private fun copyByteBuffer(buffer: ByteBuffer): ByteBuffer =
+        ByteBuffer.allocate(buffer.remaining()).put(buffer.duplicate()).flip()
+
+    private fun concatByteBuffers(parts: List<ByteBuffer>): ByteBuffer {
+        val totalSize = parts.sumOf { it.remaining() }
+        val buffer = ByteBuffer.allocate(totalSize)
+        parts.forEach { buffer.put(it.duplicate()) }
+        return buffer.flip()
+    }
+
+    override fun getObject(k: Path): CompletableFuture<ByteBuffer> {
```

Comment:
style: use expression bodies where possible (i.e. when the return value is the only thing in the method); then elide return types if it's obvious